karapace/key_format_corrector.py

```diff
@@ -0,0 +1,219 @@
+"""
+karapace - Key correction
+
+Copyright (c) 2022 Aiven Ltd
+See LICENSE for details
+"""
+
+from aiokafka import AIOKafkaConsumer, AIOKafkaProducer
+from aiokafka.structs import TopicPartition
+from asyncio import Future
+from karapace.config import Config
+from karapace.typing import JsonData
+from karapace.utils import json_encode
+from typing import List, Optional
+
+import asyncio
+import json
+import logging
+import threading
+import time
+
+LOG = logging.getLogger(__name__)
+
+SCHEMA_KEY_ORDER = ["keytype", "subject", "version", "magic"]
+CONFIG_KEY_ORDER = ["keytype", "subject", "magic"]
+NOOP_KEY_ORDER = ["keytype", "magic"]
+
+CORRECT_KEY_ORDERS = [SCHEMA_KEY_ORDER, CONFIG_KEY_ORDER, NOOP_KEY_ORDER]
+
+
+def _get_key_in_correct_order(key: JsonData) -> bytes:
+    corrected_key = {
+        "keytype": key["keytype"],
+    }
+    if "subject" in key:
+        corrected_key["subject"] = key["subject"]
+    if "version" in key:
+        corrected_key["version"] = key["version"]
+    # Magic is the last element
+    corrected_key["magic"] = key["magic"]
+    return json_encode(corrected_key, sort_keys=False, binary=True, compact=True)
+
+
+class KeyFormatCorrector:
+    """Correct message key format to use "keytype, subject, version, magic" order of keys.
+
+    Karapace has used "subject, version, magic, keytype". This can cause issues when migrating from
+    Confluent Schema Registry to Karapace. Kafka log compaction compares the key as bytes and using different
+    order of keys in the contained JSON does not produce same byte output.
+    """
+
+    def __init__(self, config: Config) -> None:
+        self.config = config
+
+        self.timeout_ms = 1000
+        self.running = False
+        self.key_correction_done = False
+        self.failed = False
+
+        self.total_msgs = 0
+        self.corrected_msgs = 0
+
+    def stop(self) -> None:
+        LOG.info("[KeyCorrector] Stopping key correction.")
+        self.running = False
+
+    def start(self) -> None:
+        if not self.running and not self.key_correction_done:
+            LOG.info("[KeyCorrector] Starting key correction.")
+            thread = threading.Thread(target=self._run, daemon=True)
+            self.running = True
+            thread.start()
+
+    def _run(self) -> None:
+        asyncio.run(self.run_correction())
+
+    async def run_correction(self) -> None:
+
+        if self.failed:
+            LOG.warning("[KeyCorrector] Correction marked failed, not attempting again.")
+            return
+
+        latest_offset = 0
+        start_time = time.monotonic()
+        producer = AIOKafkaProducer(
+            bootstrap_servers=self.config["bootstrap_uri"],
```

Comment:
This will fail with secure kafka connection (ssl and/or sasl auth) because those parameters are not passed?